okhttp/src/main/java/io/grpc/transport/okhttp/OkHttpClientStream.java

```diff
@@ -96,6 +96,10 @@ public void transportHeadersReceived(List<Header> headers, boolean endOfStream)
     synchronized (lock) {
       if (endOfStream) {
         transportTrailersReceived(Utils.convertTrailers(headers));
+        if (getTransportError() != null) {
```

Comment:
If there is no error, transport .finishStream() will be called in remoteEndClosed().
If we decide to finish a stream due to an error before seeing endOfStream, then the transport.finishStream() will be called in sendCancel().
So the only way that the stream is not cleaned up is that we finish the stream with an error after seeing endOfStream.